src/data/array/table.lean

```diff
@@ -0,0 +1,244 @@
+/-
+Copyright (c) 2018 Keeley Hoek. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Keeley Hoek
+-/
+
+import data.list
+import data.array.defs
+
+import tactic.lint
+
+/-!
+# Tables
+
+A `table` is a self-resizing array-backed list,
+which uses opaque references called `table_ref`s for access.
+-/
+
+universe variables u v w z
+
+attribute [inline] bool.decidable_eq option.is_some option.is_none list.head
+attribute [inline] array.read array.write
+
+/-- `table_ref` is a type alias for `ℕ`. -/
+@[irreducible] def table_ref : Type := ℕ
+
+namespace table_ref
+
+section internal
+
+local attribute [reducible] table_ref
+
+/-- A natural number, viewed as `table_ref`. -/
+def of_nat (r : ℕ) : table_ref := r
+
+/-- A `table_ref`, viewed as natural number. -/
+def to_nat (r : table_ref) : ℕ := r
+
+/-- The succesor function on `table_ref`s. -/
+def next (r : table_ref) : table_ref := of_nat (r + 1)
+
+lemma to_nat_next (r : table_ref) : r.next.to_nat = r.to_nat + 1 := rfl
+
+instance : decidable_eq table_ref := by apply_instance
+
+instance : inhabited table_ref := by apply_instance
+
+instance : has_well_founded table_ref := by apply_instance
+
+instance : has_to_string table_ref := by apply_instance
+
+meta instance : has_to_format table_ref := by apply_instance
+
+end internal
+
+/-- A function to convert a `table_ref` to a `string`. -/
+def to_string (r : table_ref) : string := to_string r.to_nat
+
+/-- The `table_ref` `null` is implemented as the natural number `0xFFFFFFFF`. -/
+def null  : table_ref := of_nat 0xFFFFFFFF
+
+/-- The first `table_ref`; implemented as the natural number `0`. -/
+def first : table_ref := of_nat 0
+
+end table_ref
+
+/-- An `indexed` type is a type endowed with a indexing function to `table_ref`. -/
+class indexed (α : Type u) :=
+(index : α → table_ref)
+
+/-- A type `α` is `keyed` on a type `κ` with dedicable equality,
+if `α` endowed with a keying function to `κ`. -/
+class keyed (α : Type u) (κ : Type v) [decidable_eq κ] :=
+(key : α → κ)
+
+/-- A `table` is a self-resizing array-backed list. -/
+structure table (α : Type u) :=
+(next_id : table_ref)
+(buff_len : ℕ)
+(entries : array buff_len (option α))
+
+namespace table
+
+variables {α : Type u} {β : Type v} {κ : Type w} [decidable_eq κ] (t : table α)
+
+-- TODO use push_back and pop_back builtins to avoid array preallocation
+-- TODO several recusion-induced-meta can be removed from the file (given proofs)
+
+/-- The default buffer length for tables. -/
+def DEFAULT_BUFF_LEN := 10
+
+/-- Initialize a new table, with all entries set to `none`. -/
+def create (buff_len : ℕ := DEFAULT_BUFF_LEN) : table α :=
+⟨table_ref.first, buff_len, mk_array buff_len none⟩
+
+instance : inhabited (table α) := ⟨create⟩
+
+/-- Turn a `list α` into a `table α`. -/
+def from_list (l : list α) : table α :=
```

Comment:
This function is used once for arbitrary lists, and AFAICT `to_list` is called on the result.